prusti-specs/src/extern_spec_rewriter.rs

```diff
@@ -169,20 +29,95 @@ fn rewrite_self(tokens: proc_macro2::TokenStream) -> proc_macro2::TokenStream {
     new_tokens
 }
 
-fn rewrite_method_inputs(item_ty: &syn::Type, method: &mut ImplItemMethod) ->
-    syn::punctuated::Punctuated<syn::Expr, syn::token::Comma> {
+/// Add `PhantomData` markers for each type parameter to silence errors
+/// about unused type parameters.
+///
+/// Given
+/// ```text
+/// struct Foo<A,B> {
+///    ...fields
+/// }
+/// ```
+/// Result
+/// ```text
+/// struct Foo<A,B> {
+///     ...fields
+///     ::core::marker::PhantomData<A>,
+///     ::core::marker::PhantomData<B>
+/// }
+/// ```
+fn add_phantom_data_for_generic_params(item_struct: &mut syn::ItemStruct) {
+    item_struct.generics = item_struct.generics.clone();
+    let generics = &item_struct.generics;
+
+    let mut fields: Vec<syn::Field> = Vec::new();
+    for param in generics.params.iter() {
+        let field: ParsableUnnamedField = match param {
+            syn::GenericParam::Type(tp) => {
+                let ident = tp.ident.clone();
+                syn::parse_quote! { ::core::marker::PhantomData<#ident> }
+            }
+            syn::GenericParam::Lifetime(ld) => {
+                let ident = ld.lifetime.clone();
+                syn::parse_quote! { &#ident ::core::marker::PhantomData<()> }
+            }
+            syn::GenericParam::Const(_cp) => continue,
+        };
+        // Unwrap the `ParsableUnnamedField` and push. See below.
+        fields.push(field.0);
+    }
+
+    item_struct.fields = syn::Fields::Unnamed(syn::parse_quote! { ( #(#fields),* ) });
+}
+
+/// The `syn::Field` doesn't implement `syn::parse::Parse` directly since it can do
+/// both `parse_unnamed` and `parse_named`. This is a wrapper to tell `parse_quote`
+/// to use `parse_unnamed` when parsing. https://github.com/dtolnay/syn/issues/651
+struct ParsableUnnamedField(syn::Field);
+
+impl syn::parse::Parse for ParsableUnnamedField {
+    fn parse(input: syn::parse::ParseStream<'_>) -> syn::parse::Result<Self> {
+        Ok(ParsableUnnamedField(syn::Field::parse_unnamed(input)?))
+    }
+}
+
+/// We take the Generics (parameters) defined with the `#[extern_spec] impl<...>` (the `<...>`)
+/// but then need to pass those as arguments: `SomeStruct<...>`. This function translates from
+/// the syntax of one to the other; e.g. `<T: Bound, 'l: Bound, const C: usize>` -> `<T, 'l, C>`
+pub fn rewrite_generics(gens: &syn::Generics) -> syn::AngleBracketedGenericArguments {
+    let args: Vec<syn::GenericArgument> = gens
+        .params
+        .clone()
+        .iter()
```

Comment:
If you are cloning here, you can use `.into_iter` and maybe avoid the `clone` in the `Lifetime` case?